packages/action-listener-middleware/src/index.ts

```diff
@@ -124,18 +146,165 @@ export interface CreateListenerMiddlewareOptions<ExtraArgument = unknown> {
   onError?: ListenerErrorHandler
 }
 
-export interface AddListenerAction<
-  A extends AnyAction,
-  S,
-  D extends Dispatch<AnyAction>,
-  O extends ActionListenerOptions
+interface AddListenerOverloads<
+  Return,
+  S = unknown,
+  D extends Dispatch = ThunkDispatch<S, unknown, AnyAction>
 > {
-  type: 'actionListenerMiddleware/add'
-  payload: {
-    type: string
-    listener: ActionListener<A, S, D, O>
-    options?: O
+  <MA extends AnyAction, LP extends ListenerPredicate<MA, S>>(
+    options: {
+      actionCreator?: never
+      type?: never
+      matcher?: never
+      predicate: LP
+      listener: ActionListener<ListenerPredicateGuardedActionType<LP>, S, D>
+    } & ActionListenerOptions
+  ): Return
+  <C extends TypedActionCreator<any>>(
+    options: {
+      actionCreator: C
+      type?: never
+      matcher?: never
+      predicate?: never
+      listener: ActionListener<ReturnType<C>, S, D>
+    } & ActionListenerOptions
+  ): Return
+  <T extends string>(
+    options: {
+      actionCreator?: never
+      type: T
+      matcher?: never
+      predicate?: never
+      listener: ActionListener<Action<T>, S, D>
+    } & ActionListenerOptions
+  ): Return
+  <MA extends AnyAction, M extends MatchFunction<MA>>(
+    options: {
+      actionCreator?: never
+      type?: never
+      matcher: M
+      predicate?: never
+      listener: ActionListener<GuardedType<M>, S, D>
+    } & ActionListenerOptions
+  ): Return
+
+  <LP extends AnyActionListenerPredicate<S>>(
+    options: {
+      actionCreator?: never
+      type?: never
+      matcher?: never
+      predicate: LP
+      listener: ActionListener<AnyAction, S, D>
+    } & ActionListenerOptions
+  ): Return
```

Comment:
I'll have to check on my computer to verify, but making this overload to be the first one on top should fix the state inference problem.

When I did that on my branch, I then saw a different error which I fixed using the  `ListenerPredicateGuardedActionType` I put in there.

Reply:
What I'm seeing is an action inference problem (which might be what you meant).  I just tried moving this one up, and didn't see a difference.  I did already grab `LPGuardedActionType` from your branch over to here.